src/transformers/trainer.py

```diff
@@ -233,6 +242,83 @@ def __init__(
+        optimizers: Tuple[torch.optim.Optimizer, torch.optim.lr_scheduler.LambdaLR] = (None, None),
+    ):
+        r"""
+        Creates an instance of :class:`~transformers.Trainer` from a nlp :obj:`DatasetDict` and `Metric`.
+
+        Args:
+            model (:class:`~transformers.PreTrainedModel`):
+                The model to train, evaluate or use for predictions.
+            args (:class:`~transformers.TrainingArguments`):
+                The arguments to tweak training.
+            dataset (:obj:`nlp.dataset_dict.DatasetDict`):
+                The dataset loaded and preprocessed via the :obj:`nlp` library.
+            data_collator (:obj:`DataCollator`, `optional`, defaults to :func:`~transformers.default_data_collator`):
+                The function to use to form a batch from a list of elements from the :obj:`dataset`.
+            metrics (:obj:`nlp.Metric` or :obj:`List[nlp.Metric]`, `optional`):
+                One or several metrics loaded via the :obj:`nlp` library.
+            final_activation (:obj:`str` or :class:`~transformers.trainer_utils.FinalActivation`, `optional`):
+                The final activation to apply to the model output before feeding them to the :obj:`metrics`.
+            tb_writer (:obj:`SummaryWriter`, `optional`):
+                Object to write to TensorBoard.
+            optimizers (:obj:`Tuple[torch.optim.Optimizer, torch.optim.lr_scheduler.LambdaLR`, `optional`):
+                A tuple containing the optimizer and the scheduler to use. Will default to an instance of
+                :class:`~transformers.AdamW` on your model and a scheduler given by
+                :func:`~transformers.get_linear_schedule_with_warmup` controlled by :obj:`args`.
+
+        Examples:
+
+            >>> from nlp import load_dataset, load_metric
+            >>> from transformers import AutoModelForSequenceClassification, AutoTokenizer, Trainer, TrainingArguments
+            >>> dataset = load_dataset('glue', 'sst2')
+            >>> metric = load_metric('glue', 'sst2')
+
+            >>> model_name = "distilbert-base-uncased-finetuned-sst-2-english"
+            >>> tokenizer = AutoTokenizer.from_pretrained(model_name)
+            >>> model = AutoModelForSequenceClassification.from_pretrained(model_name)
+
+            >>> encoded_dataset = dataset.map(lambda examples: tokenizer(examples['sentence'], padding=True), batched=True)
+            >>> args = TrainingArguments(output_dir = "test")
+
+            >>> trainer = Trainer.from_nlp_dataset(model, args, encoded_dataset, metrics=metric, final_activation="argmax")
+            >>> trainer.evaluate()
+            {'eval_loss': 0.3902028881816018, 'eval_accuracy': 0.9105504587155964, 'step': 0}
+        """
+        assert is_nlp_available(), "This method requires the nlp library: `pip install nlp`."
+        if metrics is not None:
+            compute_metrics = ComputeNLPMetrics(metrics, activation=final_activation)
```

Comment:
Intuitively I would say that the metric is responsible to choose which "final_activation" to use here, so I would move this into the metrics instead and think the user should not even have to specific it. Or is it very common that the same metric has multiple possible "final_activation" functions that can be applied?

Reply:
`nlp` metrics don't have an activation built in. IMO a metric is just a function, it should not have an idea of what the final activation is.
It's a task model that has an idea of the activation function, though it would not know if it needs to take argmax (for metrics that uses prediction indices like accuracy) or softmax (for metrics that use the probabilities like roc_score, auc_score). In the end, I think we could have a default that works most of the time by mapping models to a given final activation, but it still needs to be available as an argument to the user for all the cases that can go wrong.